pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,270 @@
+		"gate.players.current",
+		metric.WithDescription("Current number of connected players"),
+	)
+
+	playerConnections, err3 = meter.Int64Counter(
+		"gate.players.total",
+		metric.WithDescription("Total number of player connections"),
+	)
+
+	connDuration, err4 = meter.Float64Histogram(
+		"gate.connection.duration",
+		metric.WithDescription("Connection duration in seconds"),
+	)
+
+	for _, err := range []error{err1, err2, err3, err4} {
+		if err != nil {
+			return nil, fmt.Errorf("failed to create metrics: %w", err)
+		}
+	}
+
+	return func() {
+		if err := provider.Shutdown(ctx); err != nil {
+			fmt.Printf("failed to shutdown meter provider: %v\n", err)
+		}
+	}, nil
+}
+
+func initTracing(ctx context.Context, res *resource.Resource, cfg config.Config) (func(), error) {
+	var exporter sdktrace.SpanExporter
+	var err error
+
+	switch cfg.Telemetry.Tracing.Exporter {
+	case "otlp":
+		exporter, err = otlptracegrpc.New(ctx,
+			otlptracegrpc.WithEndpoint(cfg.Telemetry.Tracing.Endpoint),
+			otlptracegrpc.WithInsecure(),
+		)
+
+	case "stdout":
+		exporter, err = stdouttrace.New(
+			stdouttrace.WithPrettyPrint(),
+			stdouttrace.WithWriter(os.Stdout),
+		)
+
+	case "jaeger":
+		return nil, fmt.Errorf("jaeger exporter is deprecated, use OTLP exporter with jaeger collector instead")
+
+	default:
+		return nil, fmt.Errorf("unknown tracer type: %s", cfg.Telemetry.Tracing.Exporter)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to create trace exporter: %w", err)
+	}
+
+	tracerProvider := sdktrace.NewTracerProvider(
+		sdktrace.WithBatcher(exporter),
+		sdktrace.WithResource(res),
+	)
+
+	otel.SetTracerProvider(tracerProvider)
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
+		propagation.TraceContext{},
+		propagation.Baggage{},
+	))
+
+	tracer = tracerProvider.Tracer("gate")
+
+	return func() {
+		if err := tracerProvider.Shutdown(ctx); err != nil {
+			fmt.Printf("failed to shutdown tracer provider: %v\n", err)
+		}
+	}, nil
+}
+
+// RecordPlayerConnection records a new player connection metric
+func RecordPlayerConnection(ctx context.Context, username string) {
+	if playerConnections != nil {
+		playerConnections.Add(ctx, 1,
+			metric.WithAttributes(
+				attribute.String("username", username),
+			),
+		)
+	}
+}
+
+// RecordPlayerDisconnection updates metrics when a player disconnects
+func RecordPlayerDisconnection(ctx context.Context, username string, duration float64) {
+	if connDuration != nil {
+		connDuration.Record(ctx, duration,
+			metric.WithAttributes(
+				attribute.String("username", username),
+			),
+		)
+	}
+}
+
+// UpdateGathererMetrics updates observable metrics
+func UpdateGathererMetrics(_ context.Context, tps float64) {
+	tpsValue = tps // Update the TPS value that will be observed by the callback
+}
+
+// UpdatePlayerCount updates the current player count
+func UpdatePlayerCount(ctx context.Context, delta int64) {
+	if playerGauge != nil {
+		playerGauge.Add(ctx, delta)
+	}
+}
```

Comment:
These functions are never called?

and Gate has no such thing as ticks per second.

Reply:
Also globalizing these values breaks support for having multiple Gate instances in a Go program.

Reply:
Make sure to init otel where gate instance are created